pkg/controller/volume/attachdetach/metrics/metrics.go

```diff
@@ -33,58 +33,81 @@ var (
+			Name: "attachdetach_controller_forced_detaches",
+			Help: "Number of times the A/D Controller performed a forced detach"})
+)
+var registerMetrics sync.Once
 
-// Register registers pvc's in-use metrics
+// Register registers metrics in A/D Controller.
 func Register(pvcLister corelisters.PersistentVolumeClaimLister,
 	pvLister corelisters.PersistentVolumeLister,
 	podLister corelisters.PodLister,
+	asw cache.ActualStateOfWorld,
+	dsw cache.DesiredStateOfWorld,
 	pluginMgr *volume.VolumePluginMgr) {
 	registerMetrics.Do(func() {
-		prometheus.MustRegister(newVolumeInUseCollector(pvcLister, podLister, pvLister, pluginMgr))
+		prometheus.MustRegister(newVolumesCollector(pvcLister,
+			podLister,
+			pvLister,
+			asw,
+			dsw,
+			pluginMgr))
+		prometheus.MustRegister(forcedDetachMetricCounter)
 	})
+}
 
+type volumesCollector struct {
+	pvcLister       corelisters.PersistentVolumeClaimLister
+	podLister       corelisters.PodLister
+	pvLister        corelisters.PersistentVolumeLister
+	asw             cache.ActualStateOfWorld
+	dsw             cache.DesiredStateOfWorld
+	volumePluginMgr *volume.VolumePluginMgr
 }
 
-func (volumeInUse nodeVolumeCount) add(nodeName types.NodeName, pluginName string) {
-	nodeCount, ok := volumeInUse[nodeName]
+// volumeCount is a map of maps used as a counter, e.g.:
+//     node 172.168.1.100.ec2.internal has 10 EBS and 3 glusterfs PVC in use:
+//     {"172.168.1.100.ec2.internal": {"aws-ebs": 10, "glusterfs": 3}}
+//     state actual_state_of_world contains a total of 10 EBS volumes:
+//     {"actual_state_of_world": {"aws-ebs": 10}}
+type volumeCount map[string]map[string]int64
+
+func (v volumeCount) add(typeKey, counterKey string) {
+	count, ok := v[typeKey]
 	if !ok {
-		nodeCount = map[string]int{}
+		count = map[string]int64{}
 	}
-	nodeCount[pluginName]++
-	volumeInUse[nodeName] = nodeCount
+	count[counterKey]++
+	v[typeKey] = count
 }
 
-func newVolumeInUseCollector(
+func newVolumesCollector(
 	pvcLister corelisters.PersistentVolumeClaimLister,
 	podLister corelisters.PodLister,
 	pvLister corelisters.PersistentVolumeLister,
-	pluginMgr *volume.VolumePluginMgr) *volumeInUseCollector {
-	return &volumeInUseCollector{pvcLister, podLister, pvLister, pluginMgr}
+	asw cache.ActualStateOfWorld,
+	dsw cache.DesiredStateOfWorld,
+	pluginMgr *volume.VolumePluginMgr) *volumesCollector {
+	return &volumesCollector{pvcLister, podLister, pvLister, asw, dsw, pluginMgr}
 }
 
 // Check if our collector implements necessary collector interface
-var _ prometheus.Collector = &volumeInUseCollector{}
+var _ prometheus.Collector = &volumesCollector{}
 
-func (collector *volumeInUseCollector) Describe(ch chan<- *prometheus.Desc) {
+func (collector *volumesCollector) Describe(ch chan<- *prometheus.Desc) {
 	ch <- inUseVolumeMetricDesc
+	ch <- totalVolumesMetricDesc
 }
 
-func (collector *volumeInUseCollector) Collect(ch chan<- prometheus.Metric) {
+func (collector *volumesCollector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
May be we should use the name `attachDetachStateCollector` which perhaps more matches with metrics we are emitting here?

Reply:
Done.